### PR TITLE
Upgrade GetBucketRegion to aws-sdk-go-v2

### PR DESCRIPTION
**Pull request by kaovilai**
Signed-off-by: Tiger Kaovilai <tkaovila@redhat.com>

## Why the changes were made

To test that aws-sdk-go-v2 upgrade in aws plugin did not break bucket discovery to debug https://github.com/openshift/oadp-operator/pull/1451 failures.

## How to test the changes made

```
go test ./pkg/storage/aws/...
```

Followup
- https://github.com/openshift/oadp-operator/issues/1471

### CODE REVIEW

**Comment by kaovilai** [comment]
TODO: Move CloudStorageAPI for aws to aws-sdk-go-v2

**Review by mrnold — APPROVED** [review]

**Review by sseago — DISMISSED** [review]

**Comment by kaovilai** [comment]
```
--- FAIL: TestDPAReconciler_ValidateBackupStorageLocations (0.01s)
    --- FAIL: TestDPAReconciler_ValidateBackupStorageLocations/BSL_Region_not_set_for_aws_provider_without_S3ForcePathStyle_with_BackupImages_false_expect_to_succeed (0.00s)
        bsl_test.go:1503: ValidateBackupStorageLocations() error = region for AWS backupstoragelocation not automatically discoverable. Please set the region in the backupstoragelocation config, wantErr false
```

uhoh unit test

**Comment by kaovilai** [comment]
the ValidateBSL unit test' bucket does work in s3_test.go so I think might be same issue we're hitting elsewhere?

**Comment by kaovilai** [comment]
FWIW the same suite test works locally
```
❯ go test -run TestDPAReconciler_ValidateBackupStorageLocations ./controllers/...
ok      github.com/openshift/oadp-operator/controllers  1.361s
```

File: pkg/storage/aws/s3.go [hunk 1]
```diff
@@ -17,27 +18,25 @@ func BucketRegionIsDiscoverable(bucket string) bool {
...
-	sessionInstance, err := session.NewSession()
+	// GetBucketRegion will attempt to get the region for a bucket using the client's configured region to determine
+	// which AWS partition to perform the query on.
+	// Client therefore needs to be configured with region.
+	// In local dev environments, you might have ~/.aws/config that could be loaded and set with default region.
+	// In cluster/CI environment, ~/.aws/config may not be configured, so set region explicitly.
+	// Also set to use anonymous credentials
+	cfg, err := config.LoadDefaultConfig(context.TODO(),
+		config.WithRegion("us-east-1"),
+		config.WithCredentialsProvider(aws.AnonymousCredentials{}),
```
Comment (kaovilai) [inline pkg/storage/aws/s3.go:32]: This is what's needed to make this test work on environments where ~/.aws/credentials isn't specified.

**Comment by weshayutin** [comment]
glad we tested this in the scrum lolz 0/

File: pkg/storage/aws/s3.go (hunk 1 above)
Comment (mateusoliveira43) [reply pkg/storage/aws/s3.go:32]: alright, thanks

will add a hold here and added a link to velero issue in the test that broke

**Comment by mateusoliveira43** [comment]
/hold
wait velero issue be resolved

**Comment by kaovilai** [comment]
TODO: use credentials to query region on non-public buckets
https://docs.aws.amazon.com/AmazonS3/latest/API/API_HeadBucket.html#:~:text=Request%20to%20public%20buckets%20that%20grant%20the%20s3%3AListBucket%20permission%20publicly%20do%20not%20need%20to%20be%20signed.

https://github.com/aws/aws-sdk-go-v2/pull/2716#:~:text=unauthorized%20operations%20will%20only%20work%20against%20non%2Dpublic%20buckets.

**Review by weshayutin — DISMISSED** [review]
thanks @kaovilai

**Comment by kaovilai** [comment]
FYI the PR in velero is practically similar to this except that they use real credentials. I do not think it is urgent or practical to do so as far as unit test goes.. but we can implement this in the prod env as a follow up.

Benefit to real creds as opposed to anonymous is it can list bucket regions for private buckets.

**Review by shubham-pampattiwar — DISMISSED** [review]
/lgtm

File: pkg/storage/aws/s3.go [hunk 2]
```diff
@@ -17,27 +18,25 @@ func BucketRegionIsDiscoverable(bucket string) bool {
...
-
-	sessionInstance, err := session.NewSession()
+	// GetBucketRegion will attempt to get the region for a bucket using the client's configured region to determine
+	// which AWS partition to perform the query on.
+	// Client therefore needs to be configured with region.
+	// In local dev environments, you might have ~/.aws/config that could be loaded and set with default region.
+	// In cluster/CI environment, ~/.aws/config may not be configured, so set region explicitly.
+	// Also set to use anonymous credentials
+	cfg, err := config.LoadDefaultConfig(context.TODO(),
+		config.WithRegion("us-east-1"),
```
Comment (mpryc) [inline pkg/storage/aws/s3.go:31]: Can you define:
```
const DefaultAwsBucketRegion = "us-east-1"
```

File: pkg/storage/aws/s3.go [hunk 3]
```diff
@@ -17,27 +18,25 @@ func BucketRegionIsDiscoverable(bucket string) bool {
...
-			region, _ = s3manager.GetBucketRegion(context.Background(), sessionInstance, bucket, regionHint)
-
-			// we only need to try a single region hint per partition, so break after the first
-			break
-		}
-
-		if region != "" {
-			return region, nil
-		}
+	region, err = manager.GetBucketRegion(context.Background(), s3.NewFromConfig(cfg), bucket)
```
Comment (mpryc) [inline pkg/storage/aws/s3.go:37]: Don't we want to make this call first and then fall back if not gathered to default region?

File: pkg/storage/aws/s3.go
Comment (mateusoliveira43) [inline pkg/storage/aws/s3.go:1]: prior to merge, this needs to be uncommented 
https://github.com/openshift/oadp-operator/blob/5084c37f2057c94c519deca01b87f1a0ba868d6b/tests/e2e/dpa_deployment_suite_test.go#L357-L365

File: pkg/storage/aws/s3.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (kaovilai) [reply pkg/storage/aws/s3.go:37]: will just use background.. There's no plan to cancel the context.

File: pkg/storage/aws/s3.go
Comment (kaovilai) [reply pkg/storage/aws/s3.go:1]: ++

Comment (mateusoliveira43) [reply pkg/storage/aws/s3.go:1]: rebase was done, can this PR be updated to get merged?

Comment (kaovilai) [reply pkg/storage/aws/s3.go:1]: done

**Review by mateusoliveira43 — APPROVED** [review]
E2E passed, I am fine with the changes :)

did not remove `hold` label because I saw a TODO in PR description

File: pkg/storage/aws/s3.go (hunk 2 above)
Comment (kaovilai) [reply pkg/storage/aws/s3.go:31]: This isn't default. This is the to specify a hinting server. See https://github.com/aws/aws-sdk-go-v2/pull/2716#:~:text=adding%20region%20hint%20in%20the%20example%20is%20good

File: pkg/storage/aws/s3.go (hunk 3 above)
Comment (kaovilai) [reply pkg/storage/aws/s3.go:37]: This call will be made to a specific region hinting server. It cannot be called without specifying a specific region before hand.

**Comment by mateusoliveira43** [comment]
/retest
post restore flake (deployment/deploymentConfig not available)

**Pull request merged**